types/functional-red-black-tree/index.d.ts

```diff
@@ -0,0 +1,190 @@
+// Type definitions for functional-red-black-tree 1.0
+// Project: https://github.com/mikolalysenko/functional-red-black-tree
+// Definitions by: Campbell Wass <https://github.com/CamWass>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare namespace createRBTree {
+    /** Represents a functional red-black tree. */
+    interface Tree<K, V> {
+        /** Returns the root node of the tree. */
+        root: Node<K, V>;
+
+        /** A sorted array of all keys in the tree. */
+        readonly keys: K[];
+
+        /** An array of all values in the tree, sorted by key. */
+        readonly values: V[];
+
+        /** The number of items in the tree. */
+        readonly length: number;
+
+        /**
+         * Creates a new tree with the new `key/value` pair inserted.
+         *
+         * @param key The key of the item to insert.
+         * @param value The value of the item to insert.
+         * @returns A new tree with `key` and `value` inserted.
+         */
+        insert: (key: K, value: V) => Tree<K, V>;
+
+        /**
+         * Walks a visitor function over the nodes of the tree in order.
+         *
+         * @param visitor The callback to be executed on each node. If a truthy
+         * value is returned from the visitor, then iteration is stopped.
+         * @param lo An optional start of the range to visit (inclusive).
+         * @param hi An optional end of the range to visit (non-inclusive).
+         * @returns The last value returned by the callback.
+         */
+        forEach: {
+            <T>(visitor: (key: K, value: V) => T): T;
+            // tslint:disable-next-line:unified-signatures
```

Comment:
tslint gave the following error for this signiture:
`ERROR: 42:58  unified-signatures  These overloads can be combined into one signature with an optional parameter.`
From my understanding, it's recommending combining these overloads into somthing like this:
```typescript
 <T>(visitor: (key: K, value: V) => T, lo?: K, hi?: K): T;
```
However, I belive this is unsafe/not representative of the implementation, which is as follows:
```javascript
proto.forEach = function rbTreeForEach(visit, lo, hi) {
  if(!this.root) {
    return
  }
  switch(arguments.length) {
    case 1:
      return doVisitFull(visit, this.root)
    break

    case 2:
      return doVisitHalf(lo, this._compare, visit, this.root)
    break

    case 3:
      if(this._compare(lo, hi) >= 0) {
        return
      }
      return doVisit(lo, hi, this._compare, visit, this.root)
    break
  }
}
```
To me, because the function uses `arguments.length`, it looks like it can take:

- Just a visitor
- A visitor and a lower bound of type `K`
- A visitor, lower bound of type `K`, and a upper bound of type `K`

The combined signiture (with it's two optional parameters) however, allows the following:
```javascript
tree.forEach(
  // visitor:
  () => {}, // ok
  // lo:
  undefined, // potentially unexpected/unsafe
  // hi:
  5 // ok (assuming the type "K" is "number")
)
```

Is it acceptable to disable the lint rule for this line, or is there a combined signiture that correctly matches the implementation that I'm missing?